### PR TITLE
wsutil: add MaxFrameSize to Reader

### PR DESCRIPTION
**Pull request by agnivade**
Fixes #26

### CODE REVIEW

File: wsutil/reader.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ import (
 // preceding NextFrame() call.
 var ErrNoFrameAdvance = errors.New("no frame advance")
 
+var ErrReadLimit = errors.New("read limit exceeded")
```
Comment (cristaloleg) [inline wsutil/reader.go:15]: Can you add a comment here? So `golint` will not yell at this line.

File: wsutil/reader_test.go [hunk 2]
```diff
@@ -168,6 +168,32 @@ func TestReaderNextFrameAndReadEOF(t *testing.T) {
 
 }
 
+func TestMaxFrameSize(t *testing.T) {
+	var buf bytes.Buffer
+	f := ws.NewTextFrame([]byte("small frame")) // 11 bytes
```
Comment (cristaloleg) [inline wsutil/reader_test.go:173]: what about to extract 11 or 10 to the `const` or just a `var` ?

Comment (agnivade) [reply wsutil/reader_test.go:173]: I have a better idea ;)

Comment (cristaloleg) [reply wsutil/reader_test.go]: Yep, exactly what the next commits are :)

File: wsutil/reader.go (hunk 1 above)
Comment (gobwas) [reply wsutil/reader.go:17]: I would suggest to rename the error to `ErrFrameTooLarge` to be more connected in the Reader.MaxFrameSize.

File: wsutil/reader.go [hunk 3]
```diff
@@ -42,7 +46,12 @@ type Reader struct {
 	// header RSV segment.
 	Extensions []RecvExtension
 
-	// TODO(gobwas): add max frame size limit here.
+	// MaxFrameSizeBytes controls the maximum frame size in bytes
+	// that can be read. A message exceeding that size will return
+	// a ErrReadLimit to the application.
+	//
+	// Not setting this field means there is no limit.
+	MaxFrameSizeBytes int64
```
Comment (gobwas) [inline wsutil/reader.go:54]: Can we drop the `Bytes` suffix? I think it's implies bytes without any other explicit specification?

File: wsutil/reader.go [hunk 4]
```diff
@@ -175,6 +184,10 @@ func (r *Reader) NextFrame() (hdr ws.Header, err error) {
 		return hdr, err
 	}
 
+	if r.MaxFrameSizeBytes > 0 && r.MaxFrameSizeBytes < hdr.Length {
```
Comment (gobwas) [inline wsutil/reader.go:187]:
```suggestion
	if n := r.MaxFrameSizeBytes; n > 0 && hdr.Length > n {
```

**Review by gobwas — CHANGES_REQUESTED** [review]

**Comment by gobwas** [comment]
Thank you! It's great to have the test on this!

File: wsutil/reader.go (hunk 3 above)
Comment (agnivade) [reply wsutil/reader.go:54]: I don't think it hurts to be explicit. But I'm okay if you prefer to drop `Bytes` as long as we mention that in the godoc.

Comment (gobwas) [reply wsutil/reader.go:54]: Yeah, lets drop it. 👍🏻

**Review by cristaloleg — APPROVED** [review]

File: wsutil/reader.go [hunk 5]
```diff
@@ -12,6 +12,10 @@ import (
 // preceding NextFrame() call.
 var ErrNoFrameAdvance = errors.New("no frame advance")
 
+// ErrFrameTooLarge indicates that a message of length higher than
+// MaxFrameSize was being read.
+var ErrFrameTooLarge = errors.New("read limit exceeded")
```
Comment (gobwas) [inline wsutil/reader.go:17]:
```suggestion
var ErrFrameTooLarge = errors.New("frame too large")
```

**Comment by gobwas** [comment]
@agnivade Thanks! Are you okay if I squash&merge this?

**Review by gobwas — APPROVED** [review]

**Review by cristaloleg — APPROVED** [review]

**Comment by agnivade** [comment]
@gobwas - Yep, go ahead. All yours!

**Comment by agnivade** [comment]
@gobwas - Would it be possible to have a release cut for this ? It will be great not to have to pin to a commit.

**Comment by gobwas** [comment]
Hi @agnivade, sure, I did it already yesterday. It's pre-release v1.1.0-rc.2, since we have rc version of compression in master already. I will bump it as stable in one week or something like that.

**Comment by agnivade** [comment]
Great, thanks!

**Pull request merged**